### ISSUE TITLE
Remaining game length as auxiliary prediction target

### ISSUE THREAD

**Issue by TFiFiE** [lightvector/KataGo#38]
Did you consider or even already experiment with having the network predict the remaining number of moves in a game? Game length is admittedly kind of a dubious concept in the game of go (depending on the ruleset), but it's otherwise universal enough for it to be nice to find out whether predicting it might prove (significantly) beneficial. As a bonus, the info would be valuable for use in the time management (especially if the predictions are further divided by outcome (win/loss/draw)).

**Comment by isty2e** [comment]
I don't think this is a well-defined and sound target, since the game can progress indefinitely without loss of score, and is largely dependent on the number of ko remaining. Might make sense in the Japanese ruleset, however it is still dubious in terms of strength gain nonetheless. Remaining game length seems like a more valid target for chess rather than go.

**Comment by poptangtwe** [comment]
Chess

**Comment by lightvector** [comment]
Thanks for the suggestion! Yes, I've considered such a target, but I don't think it would overall be that interesting in Go. It would definitely be more interesting in games that behave more "sudden-death"-like.

I don't think you even want this prediction for time management, because the endgame can vary quite a bit in length, but unless the game is close to within a single point or two, you can play the endgame fairly quickly regardless of how long it is, meaning that a decent fraction of the time a lot of the variation in this metric is just noise in terms of how you should spend your time.

A more useful value for time management might be the following: 
![CodeCogsEqn](https://user-images.githubusercontent.com/11942395/61178091-8cb65780-a5b3-11e9-9309-ebcfa8b99b28.gif)
where V(t) is the MCTS winrate on turn t, and t_0 is the current turn number. That is, you would be predicting the time over which the variance in the game outcome will arrive, summing across all of the game's remaining variance. A low value would indicate either that the game is already almost decided or that the game is about to be decided in the next few moves, such in a uncertain but huge fight that will resolve shortly. A high value would indicate that there is a long game remaining before the outcome is likely to become certain.

Actually, this seems like a pretty neat value to predict! Whereas the number of turns is not likely to be useful for Go, maybe this value would have some value as an auxillary target (and is also completely general across games). I might try this out at some point.

**Comment by lightvector** [comment]
As a point of reference/intuition, if the MCTS winrate is a [martingale](https://en.wikipedia.org/wiki/Martingale_(probability_theory)), then the following slightly different sum, not weighting by turn number:

![CodeCogsEqn2](https://user-images.githubusercontent.com/11942395/61178218-7a89e880-a5b6-11e9-87c5-64a11143b626.gif)

would be precisely equal to:

![CodeCogsEqn3](https://user-images.githubusercontent.com/11942395/61178224-8c6b8b80-a5b6-11e9-871f-43bf477141ff.gif)

i.e. the expected amount of variance remaining is entirely determined by the current probability of winning. So weighting by time simply asks in expectation how far this fixed expected amount of variance will occur. MCTS winrate is definitely not a martingale, but an appropriate monotone function of it probably would be very close to one.

**Comment by poptangtwe** [comment]
@lightvector 
How about introducing opponent's miss rate and the loss of score as the other auxiliary prediction targets to improve handicap games?

**Comment by Splee99** [comment]
Instead of waiting for the opponent to make mistake, it is more important to create tricky game states to invite opponent's mistake. I guess in the term of MCTS, the goal maybe to build a wide search tree and choose the move with wide standard deviation of winrate (instead of LCB). However, special training may also be needed for this structure.
